src/backend/question/serializers.py

```diff
@@ -36,11 +43,70 @@ def get_question_des(self, obj):
     def get_category_name(self, obj):
         return obj.category.name
 
+    def get_suggestion_level(self, solutionObj, userObj, paramObj={
+        "w1": 1, "w2": 1, "w3": 1, "w4": 1,
+        "is_review_mode": False,
+        "user_frequency": 0
+    }):
+        '''        
+        :solutionObj 
+            current solution object
+        :userObj 
+            user object that requested this event 
+
+        ## ALL CONSTANTS STARTS WITH `CONSTANT_`   
+        '''
+        try:
+            # AL
+            # Define the adjust level here
+            CONSTANT_AL = 5
+            user_ability = np.asanyarray(userObj.ability_id)
+            solution_ability = np.asanyarray(solutionObj.ability_id)
+            AL = ((user_ability + CONSTANT_AL *
+                   (np.ones(user_ability.shape))) - solution_ability)
+
+            # CW
+            # Getting user submission for this solution, ordered by time created DESC
+            CONSTANT_CW = 1
+            userSubmissions = UserSubmission.objects.filter(
+                user_id=userObj.id, solution_id=solutionObj.id).order_by('-created_at')
+            if userSubmissions.count() == 0:
+                CW = 0
+            else:
+                cw = userSubmissions.first().details['score']
+                CW = int(cw) * CONSTANT_CW if cw is not None else 0
+
+            # SW
+            CONSTANT_SW_1 = 1
+            CONSTANT_SW_2 = 0.3
+            last_submission_date = userSubmissions.first().created_at.date() if (
+                userSubmissions.count() > 0) else float('inf')
+            date_elapsed = abs(
+                (datetime.now().date() - last_submission_date).days)
+            SW = CONSTANT_SW_1 * (1-np.exp(-CONSTANT_SW_2 * date_elapsed))
+
+            # RW
+            CONSTANT_RW = 1
+            if paramObj['is_review_mode']:
+                RW = 0
+            else:
+                RW = CONSTANT_RW * \
+                    paramObj['user_frequency'] * \
+                    (1 if userSubmissions.count() > 0 else 0)
+
+            SL = - paramObj['w1'] * AL - paramObj['w2'] * CW \
+                + paramObj['w3'] * SW + paramObj['w4'] * RW
+            return SL
+
+        except Exception as e:
```

Comment:
Use a big try except code block would be hard to debug and maintain, we can use [Exception](https://www.django-rest-framework.org/api-guide/exceptions/) like:

    try:
        code_block
    except IndexError:
        raise AuthenticationFailed('message you want to return')
    except other error:
        raise PermissionDenied('message you want to return')
    else:
         if a == b:
             raise NotFound('message you want to return')